### DISCUSSION TITLE
Group items with bracket doesn't work

### DISCUSSION THREAD

**Discussion by romado77** [FelixKratz/SketchyBar#482] [Q&A] [answered]
Hi all,

I'm currently working on a tmux plugin and have encountered an issue with bracket element. I'm seeking guidance on what I might be doing wrong.

My layout is following:
tmux.session | tmux (seprator) | tmux.window.1 | tmux.window.2 | tmux.window.n

<img width="719" alt="image" src="https://github.com/FelixKratz/SketchyBar/assets/17646835/7762d4b5-4891-4b16-9216-cfab2c5e075a">

I'm attempting to group tmux items using brackets.

```
sketchybar --add bracket status tmux.session tmux '/tmux\.window\..*/' \
           --set status "${status_bracket[@]}"
```
<img width="458" alt="image" src="https://github.com/FelixKratz/SketchyBar/assets/17646835/5a61af81-64d1-4eb0-a8cb-13489ff851d3">

On screenshot above, I enabled white border for bracket, so it's stretches to the first tmux.window element and stops there.

When I query the bracket element, it only displays two included items
<img width="293" alt="image" src="https://github.com/FelixKratz/SketchyBar/assets/17646835/f7a39524-3e3f-49ac-a829-6ea1500ab92b">

When I query the first and subsequent tmux.session elements, they are displayed, but they are not included within the bracket.

<img width="438" alt="Cursor_and_tmux_attach_-t_drv" src="https://github.com/FelixKratz/SketchyBar/assets/17646835/d5b9a7d0-9552-4270-a89e-ff283e80e93c">

I attempted to add the item 'tmux.window.1' instead of using regex, but unfortunately, this approach also did not work.

I would greatly appreciate it if someone could guide me in the right direction.

I'm on Sonoma 14.1.2 with Rectangle.

Configuration files:
### items/tmux.sh
```
#!/usr/bin/env zsh

status_bracket=(
  # background.color=0xffed8796
  background.border_color=0xffffffff
  background.border_width=2
)

sketchybar --add item tmux.session left \
    --set tmux.session \
    icon=􀐜  \
    icon.color=0xffb4befe \
    icon.font="$FONT_FACE:Bold:15.0" \
    icon.padding_left=10 \
    icon.padding_right=10 \
    label.padding_right=10 \
    # background.color=0xff585b70

sketchybar --add event tmux_window_changed \
    --add item tmux left \
    --set tmux \
    background.color=0x00000000 \
    script="$PLUGIN_DIR/tmux.sh" \
    --subscribe tmux tmux_window_changed

# sketchybar --add bracket status tmux.session tmux '/tmux\.window\..*/' \
#            --set status "${status_bracket[@]}"
```

### plugins/tmux.sh
```
#!/usr/bin/env zsh

WINDOW_DEFAULT_COLOR=0xffe5c07b
TMUX_SESSION=$(tmux ls | cut -d':' -f1) 

if [[ -z $TMUX_SESSION ]]; then
    return
fi

sketchybar --remove '/tmux_window.*/'

TMUX_WINDOWS=$(tmux list-windows -t ${TMUX_SESSION} -F '#I "#W" #{window_active}')

echo "$TMUX_WINDOWS" | while IFS= read -r line; do
    id=$(echo "$line" | awk '{print $1}')
    name=$(echo "$line" | awk '{print $2}' | tr -d '[:punct:]')
    active=$(echo "$line" | awk '{print $3}')

    if [[ $active -eq 1 ]]; then
        WINDOW_COLOR=0xffed8796
    else
        WINDOW_COLOR=$WINDOW_DEFAULT_COLOR
    fi

    sketchybar --add item tmux.window.$id left \
        --set tmux.window.$id \
        label="$name" icon="$id" \
        icon.padding_left=8 \
        icon.padding_right=8 \
        icon.font="$FONT_FACE:Semibold:12.0" \
        label.color=$WINDOW_COLOR \
        label.font="$FONT_FACE:Semibold:12.0" \
        label.padding_right=10
done

sketchybar --set tmux.session label="$TMUX_SESSION"
```

### skethchybarrc
```
#!/usr/bin/env zsh

SKETCHYBAR_CONFIG="$HOME/.config/sketchybar"
ITEM_DIR="$HOME/.config/sketchybar/items"
PLUGIN_DIR="$HOME/.config/sketchybar/plugins"

FONT_FACE="JetBrainsMono Nerd Font"

sketchybar --bar \
    height=32 \
    color=0x00000000 \
    margin=0 \
    sticky=on \
    padding_left=23 \
    padding_right=23 \
    notch_width=188 \
    display=main

sketchybar --default \
    background.color=0x66494d64 \
    background.corner_radius=5 \
    background.padding_right=5 \
    background.height=26 \
    icon.font="$FONT_FACE:Medium:15.0" \
    icon.padding_left=5 \
    icon.padding_right=5 \
    label.font="$FONT_FACE:Medium:12.0" \
    label.color=0xffcad3f5 \
    label.y_offset=0 \
    label.padding_left=0 \
    label.padding_right=5

source "$ITEM_DIR/space.sh"
source "$ITEM_DIR/tmux.sh"
source "$ITEM_DIR/front_app.sh"
source "$ITEM_DIR/clock.sh"
source "$ITEM_DIR/battery.sh"
source "$ITEM_DIR/volume.sh"
source "$ITEM_DIR/keyboard.sh"

sketchybar --update
sketchybar --trigger space_change

```

**Comment by FelixKratz** [comment, marked as answer]
I think the problem is that the tmux.window items are being created after the bracket is created. If the items do not exist when creating the bracket, they will not be included. You need to remove and readd the bracket in your plugins/tmux.sh script after adding the tmux.window items then it should work.

Alternatively, instead of deleting and readding the tmux.window items in the plugins/tmux.sh script, you could add some number (say 16) of tmux.window items in the items/tmux.sh already and simply toggle them visible/invisible with the `drawing` property in the tmux script, then your bracket logic should also work fine.

**Reply by romado77** [reply]
Indeed, moving bracket into plugin did the trick.
<img width="685" alt="image" src="https://github.com/FelixKratz/SketchyBar/assets/17646835/f3e8d803-e6b1-41c7-8641-cff54e2bd9b2">

I need to work on styling though.

**Reply by johnMocingbird** [reply]
@romado77  This looks great! Would you mind sharing your most up-to-date version of this?
